### DISCUSSION TITLE
need to hack code39 behaviour

### DISCUSSION THREAD

**Discussion by abubelinha** [py-pdf/fpdf2#971] [Ideas]
I see [this thread](https://github.com/py-pdf/fpdf2/discussions/433) is the only discussion about [code39 barcodes](https://py-pdf.github.io/fpdf2/Barcodes.html#code-39).

In the fpdf-based php script I am trying to port to fpdf2-python, there was a very tiny modification of [Code39()](http://www.fpdf.org/en/script/script46.php).
So that I could choose the line proportions in generated barcodes (I can't recall exactly how I decided to do this but I think that made barcodes easier to read with my mobile, at least for a particular label size I needed to print some time ago).  

This was a hack made by me and I think the equivalent change in python would only affect the beginning of code39() definition:
https://github.com/py-pdf/fpdf2/blob/aa3b1f4a1e16b1d513e5064278b2fb2f5b2747af/fpdf/fpdf.py#L4517C1-L4519C35

**Current definition:**
```Python
def code39(self, text, x, y, w=1.5, h=5):
        """Barcode 3of9"""
        dim = {"w": w, "n": w / 3}
```

**Modified version:**
```Python
def code39(self, text, x, y, w=1.5, h=5, proportion=3):
        """Barcode 3of9"""
        dim = {"w": w, "n": w / proportion}
```

I have very low Python programming skills but I think this change wouldn't break compatibility with previous versions.
I doubt this hack can be useful for someone else (I don't see myself improving barcodes technology), but I leave that to you as a possible think to change.

Anyway, I need to do this hack in order to port my php script, so I ask your opinion about the most correct/easiest way to do this in Python.

How can I do this right in my script, without touching the installed `fpdf2` package? 
So when I upgrade `fpdf2` version, my script can keep working and using my modified `code39()` version.
(at risk of missing only other possible improvements of that specific `code39()` function which `fpdf2` can make in the future) 

Thanks a lot in advance for your suggestions
@abubelinha

**Comment by gmischler** [comment]
I don't have access to ISO/IEC 16388, but all the descriptions of code39 I can find say that the ratio should be between 2:1 and 3:1 (recommended), so I suspect that is exactly what the standard says as well.
There's not really a point in adding a feature to the library for the sole purpose of creating non-standard barcodes.

Of course you're free to subclass FPDF and overwrite the `.code39()` method for your own purposes. The code you show above should indeed do the trick.
But then, maybe you want to run some new tests first? Even if a phone camera from many years ago may have had difficulties reading a standards-conforming barcode, that will likely not be the case anymore today. Figuring that out might save you from the hassle of creating and maintaining a custom version of fpdf2. Who cares if your barcode is a few mm shorter from now on?

**Comment by abubelinha** [comment]
> Of course you're free to subclass FPDF and overwrite the .code39() method for your own purposes. The code you show above should indeed do the trick.

Yes, that's what I want to do (implementing also the standard ratio limits you told me about).

I'd like to apply the changes within my script, without touching the installed `fpdf2` module? So that ...
- if I install other people scripts, their `pdf.code39()` calls keep working
- if my script calls `pdf.code39()`, it accepts my new `ratio` optional parameter 
- when I upgrade fpdf2 in the future, all the above keeps working (I shouldn't loose any new `fpdf2` features since its `code39` function is not expected to be modified, as per your comments)

I guess something like this is the most proper way to do it?

**fpdf_modified.py:**
```Python
from fpdf import FPDF
#from fpdf import check_page # @check_page is a decorator of code39() in FPDF. How do I use it here?

#@check_page   # commented because the above import  doesn't work
def _new_code39(self, text, x, y, w=1.5, h=5, ratio=3):
	"""Barcode 3of9 (modified with non-forced, optional user-defined ratio)"""
	if ratio>3 or ratio<2: # https://en.wikipedia.org/wiki/Code_39
		ratio=3 # use our default to keep within the standard
	dim = {"w": w, "n": w / ratio}
	(...)

FPDF.code39 = _new_code39 # this rewrites the original function so myscript.py can keep using its name
````
**myscript.py:**
```Python
from fpdf_modified import FPDF # imported from the above script
pdf = FPDF(orientation="P", unit="mm", format="A4")
pdf.add_page()
pdf.code39("*fpdf modified code39*", x=30, y=50, w=1.5, h=20, ratio=2.8)
pdf.output("fpdf-hacked-code39-test.pdf")
```

Although the above code seems to work, I still have some questions:
1. The original `code39()` function has a `@check_page` decorator.  What's the proper way of importing/implementing it in my `fpdf_modfied.py` script?  I'm concerned about which things could go wrong if I don't use it.
2. The [cod39 example](https://py-pdf.github.io/fpdf2/Barcodes.html#code-39) in fpdf2 documentation shows an Arial text below the barcode (so a human can actually read the info in the code). It was like that also in the original [PHP Code39() extension](http://www.fpdf.org/en/script/script46.php).  
But current `code39()` method does not output that text. Do you know if it was present in any of the previous Python versions of `fpdf` and later removed, so I don't need to go back to PHP code and figure out how to do that? (I might want to make human-legible codes too as  boolean option in my versioned function ... as long as that doesn't break compatibility with other fpdf2 features, like new `rotation` method).

Thanks a lot again for your time

**Reply by gmischler** [reply]
Using inheritance and overwriting the method is much simpler than injecting it into the original.

```python
class MyFPDF(FPDF):
    @check_page
    def code39(self, text, x, y, w=1.5, h=5, ratio=3):
        ...

pdf = MyFPDF()
```
The `@check_page` decorator simply checks if you already have a page added to the document, since executing the method doesn't make any sense otherwise.

I don't think the Python version ever added the text below the barcode (which means the graphic example in the documentation is incorrect). That should be easy enough to add to your version if you need it.

**Reply by abubelinha** [reply]
Thanks for the advice @gmischler 
Sorry for the dummy questions. 
To have the modified class available in a separate `fpdf_modified.py` file as above, is this a correct way to do it?
(it works, except by @check_page ... but as you put it in your code I guess I am doing something wrong):

**fpdf_modified.py:**
```Python
import fpdf
class MyFPDF(FPDF): 
	# Should I redefine check_page() in this file, outside the class?
	#@check_page # still NOT working (error if I import/invoque it).
	def code39(self, text, x, y, w=1.5, h=5, ratio=3):
		...
```
**myscript.py:**

```Python
from fpdf_modified import MyFPDF # imported from the above script
pdf = MyFPDF(orientation="P", unit="mm", format="A4")
pdf.add_page()
pdf.code39("*fpdf modified code39*", x=30, y=50, w=1.5, h=20, ratio=2.8)
pdf.output("fpdf-hacked-code39-test.pdf")
```

**Reply by gmischler** [reply]
> it works, except by @check_page

Since we're using it in a different file, that needs to be a qualified reference: `@fpdf.check_page`
I didn't actually test my example code before posting it... :wink:

**Reply by abubelinha** [reply]
Thanks. I also posted a wrong line in my last **fpdf_modified.py**. I was actually using it like this:
 `from fpdf import FPDF` 
... instead of 
`import fpdf`

But I must be missing something else. Using both `import` variants, calling `@fpdf.check_page` returns an error:

**fpdf_modified.py:**
```Python
	import fpdf # added trying to have access to check_page()
	from fpdf import FPDF
	class MyFPDF(FPDF): 
		@fpdf.check_page
		def code39(self, text, x, y, w=1.5, h=5, fontsize=10, ratio=3):
...
```
When running **myscript.py** I get: `AttributeError: module 'fpdf' has no attribute 'check_page'`

Looks like even with `import fpdf` is not possible to access [`check_page()`](https://github.com/py-pdf/fpdf2/blob/aa3b1f4a1e16b1d513e5064278b2fb2f5b2747af/fpdf/fpdf.py#L215C20-L224) function from **fpdf_modified.py**?
I don't understand why.

Tried also `from fpdf import check_page` but it doesn't work either:
`ImportError: cannot import name 'check_page' from 'fpdf'`

**Comment by gmischler** [comment]
Oops, I just realized that I've been leading you on the wrong path with `check_page`.
As it happens, it is not included in the `__all__` list of fpdf.py, so you can't import it into another file.

So the simples way is to just leave it away. After all, its only purpose is to give you an easier to understand error message if you try to add some content to a document with no pages added yet.

**Comment by abubelinha** [comment]
No problem.  With this discussion I got examples on how to do things in Python.

Thanks for all your detailed explanations!
